### PR TITLE
Add support for special tokens in nvtext::subword_tokenizer

### LINKED ISSUE #6937: [FEA] Support [CLS] [SEP] for subword_tokenize to handle correctly
Hi @VibhuJawa ,

As we discussed in the chat, it seems there was an problem that `subword_tokenize` would not handle special tokens (e.g., `[CLS]`, `[SEP]`) correctly, and I'm creating this github issue to track it.

Thanks!
Shang

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #6937 
This PR adds support for the following 7 special tokens in the `subword_tokenize`
[BOS], [EOS], [UNK], [SEP], [PAD], [CLS], and [MASK]
Descriptions for these can be found in links/text found in #6937

These can be placed anywhere in the text and may be upper or lower-case. They will be recognized regardless if they exist in the given vocabulary hash table. Example using vocab-hash.txt and code snippet from #6937
```
>>> text = '[CLS]I ate dinner.[SEP][BOS]It was yummy.[EOS]'
>>> cudf_ser = cudf.Series([text])
>>> tokens, attention_masks, metadata = cudf_ser.str.subword_tokenize('vocab-hash.txt', do_lower=True, do_truncate=False)
>>> print(id2vocab[tokens[0:17].get()])
['[CLS]' 'i' 'ate' 'dinner' '.' '[SEP]' '[BOS]' 'it' 'was' 'yummy' '.'
 '[EOS]' '[PAD]' '[PAD]' '[PAD]' '[PAD]' '[PAD]']
```

A new gtest was added for this feature.
This requires no API change to the C++ or Python interfaces.

### CODE REVIEW

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

**Review by devavret — APPROVED** [review]

File: cpp/src/text/subword/wordpiece_tokenizer.cu [hunk 1]
```diff
@@ -141,6 +142,125 @@ __global__ void mark_string_start_and_ends(uint32_t const* code_points,
...
+                 end_word_indices + end_index + 1,
+                 std::numeric_limits<uint32_t>::max());
+
+    // reset the new end-word index
+    end_word_indices[end_pos] = end_pos + 1;
+  }
+
+  uint32_t* code_points;
+  uint32_t* start_word_indices;
+  uint32_t* end_word_indices;
```
Comment (karthikeyann) [inline cpp/src/text/subword/wordpiece_tokenizer.cu:260]:
```suggestion
  uint32_t* const code_points;
  uint32_t* const start_word_indices;
  uint32_t* const end_word_indices;
```

File: cpp/src/text/subword/wordpiece_tokenizer.cu [hunk 2]
```diff
@@ -345,6 +465,14 @@ void wordpiece_tokenizer::tokenize(uvector_pair& cps_and_offsets, rmm::cuda_stre
                                                          num_strings);
   CHECK_CUDA(stream.value());
 
+  // check for special tokens and adjust indices
+  thrust::for_each_n(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<uint32_t>(0),
```
Comment (karthikeyann) [inline cpp/src/text/subword/wordpiece_tokenizer.cu:471]:
```suggestion
    thrust::make_counting_iterator<size_t>(0),
```
or `mark_special_tokens::operator()(uint32_t idx)`

File: cpp/src/text/subword/wordpiece_tokenizer.cu [hunk 3]
```diff
@@ -141,6 +142,125 @@ __global__ void mark_string_start_and_ends(uint32_t const* code_points,
...
+    if (!is_special_token(code_points + start_pos, size)) return;
+
+    // special token found
+    // adjust code-points
+    auto const end_pos = end_index - 2;
+    // change _[_ttt_]_ to _[TTT]_
+    for (auto left_idx = start_pos - 1; left_idx <= end_pos; ++left_idx) {
+      auto const cp         = code_points[left_idx + 1];
+      code_points[left_idx] = cp >= 'a' ? cp - 'a' + 'A' : cp;
+    }
```
Comment (karthikeyann) [inline cpp/src/text/subword/wordpiece_tokenizer.cu:241]: could this be replaced with transform? (if not very readable, skip it)
```suggestion
    thrust::transform(thrust::seq, 
                      code_points + start_pos,
                      code_points + end_pos + 2,
                      code_points + start_pos - 1,
                      [](auto const cp) {
                          return cp >= 'a' ? cp - 'a' + 'A' : cp; 
                      });
```

**Review by karthikeyann — APPROVED** [review]
LGTM.
minor type suggestions.

File: cpp/src/text/subword/wordpiece_tokenizer.cu (hunk 3 above)
Comment (davidwendt) [reply cpp/src/text/subword/wordpiece_tokenizer.cu:241]: This would be better as a kind of `rotate` but with a transform twist.  I don't really like `transform` here for this since input/output overlap which violates the precondition for `thrust::rotate` as documented here: https://thrust.github.io/doc/group__transformations_ga281b2e453bfa53807eda1d71614fb504.html#ga281b2e453bfa53807eda1d71614fb504

**Comment by davidwendt** [comment]
@gpucibot merge

**Comment by ttnghia** [comment]
Hi there!

I got this compilation error due to this PR:

```
.../cudf/cpp/src/text/subword/load_hash_file.cu:128:25: error: catching polymorphic type ‘class std::exception’ by value [-Werror=catch-value=]
```
